hera_cal/reflections.py

```diff
@@ -267,25 +239,445 @@ def write_reflections(self, output_calfits, input_calfits=None, overwrite=False,
+                keys that will be used as separate basis functions
+                to project onto the u modes. Example: for a given
+                cross-corr key, you can provide both auto-corr keys.
+            u : DataContainer
+                Object to pull u modes from. Default is self.umodes
+            flags : DataContainer
+                Object to pull data flags from.
+            index : int
+                Index of the u-mode to project auto-correlation onto.
+                All other u-mode indices are copied as-is into umode_interp
+                This should almost always be set to zero.
+            auto_delay : float
+                Delay in nanosec of autocorrelation to project onto the u-mode.
+                This should almost always be set to zero.
+            overwrite : bool
+                If True, overwrite output data if it exists.
+            verbose : bool
+                If True, report feedback to stdout.
+        """
+        # type check
+        assert len(keys) == len(auto_keys), "len(keys) must equal len(auto_keys)"
+
+        # get u and flags
+        if u is None:
+            u = self.umodes
+        if flags is None:
+            flags = DataContainer(dict([(k, np.zeros_like(self.pcomps[k][0], dtype=np.bool)) for k in u]))
+
+        # get dly index
+        select = np.argmin(np.abs(self.delays - auto_delay))
+
+        if not hasattr(self, 'umode_interp'):
+            self.umode_interp = DataContainer({})
+            self.umode_interp.times = u.times
+        if not hasattr(self, 'umode_interp_flags'):
+            self.umode_interp_flags = DataContainer({})
+            self.umode_interp_flags.times = u.times
+
+        # iterate over keys
+        for k, ak in zip(keys, auto_keys):
+            # check overwrite
+            if k in self.umode_interp and not overwrite:
+                echo("{} exists in umode_interp and overwrite == False, skipping...".format(k), verbose=verbose)
+                continue
+
+            if k not in u:
+                echo("{} not in u container, skipping...".format(k, ak), verbose=verbose)
+                continue
+
+            # get u-mode
+            _u = u[k][:, index]
+
+            # get autocorr
+            if isinstance(ak, list):
+                if not np.all([_ak in self.dfft for _ak in ak]):
+                    echo("{} not in self.dfft, skipping...".format(ak), verbose=verbose)
+                    continue
+                A = np.array([self.dfft[_ak][:, select] for _ak in ak]).T
+                af = np.sum([flags[_ak] for _ak in ak], axis=0)
+            else:
+                if ak not in self.dfft:
+                    echo("{} not in self.dfft, skipping...".format(ak), verbose=verbose)
+                    continue
+                A = np.array([self.dfft[ak][:, select]]).T
+                af = flags[ak]
 
-def _form_gains(epsilon):
+            # form least squares estimate
```

Comment:
You need to explain the reasoning behind this step. What's going on here and why?